pkg/common/channel.go

```diff
@@ -0,0 +1,70 @@
+package common
+
+import "sync/atomic"
+
+// In Go, unbounded channel means something different than what it means in Rust.
+// I.e. unlike Rust, "unbounded" in Go means that the channel has **no buffer**,
+// meaning that each attempt to send will block the channel until the receiver
+// reads it. Majority of primitives here in `waterfall` are designed under assumption
+// that sending is not blocking.
+const UnboundedChannelSize = 128
+
+// Creates a new channel, returns two counterparts of it where one can only send and another can only receive.
+// Unlike traditional Go channels, these allow the receiver to mark the channel as closed which would then fail
+// to send any messages to the channel over `Send“.
+func NewChannel[M any]() (Sender[M], Receiver[M]) {
+	channel := make(chan M, UnboundedChannelSize)
+	closed := &atomic.Bool{}
+	sender := Sender[M]{channel, closed}
+	receiver := Receiver[M]{channel, closed}
+	return sender, receiver
+}
+
+// Sender counterpart of the channel.
+type Sender[M any] struct {
+	// The channel itself.
+	channel chan<- M
+	// Atomic variable that indicates whether the channel is closed.
+	receiverClosed *atomic.Bool
```

Comment:
My understanding is that this is considered a bit of a code smell in Go, since a sender should never be trying to write to a closed channel. I guess this means we have something where we can't know this and need to check each time?

Reply:
Good that you've noticed it: tbh I did not like the part of writing this particular structure as it felt like it's not very elegant, but unfortunately I did not find a better way to write it.

The problem that I'm trying to solve is to indeed not write to the closed channel, **but** the problem is that from Go's standpoint the channel is not closed since the channels could only be closed in Go from the sender's side, but not from the receiver's side.

A practical example where it happens in our code:
- There is a router in the code: a structure that simply has a map of all conferences running on the SFU. The router receives signaling messages (To-Device messages), checks their `conf_id,` and simply sends the message to the conference that is responsible for this message. So there is a channel between the router and a particular conference.
- Conference is listening to the channel and reacting to the messages. But at some point, all participants leave the conference which means that the conference must be ended. So the conference grouting and main loop end, sending a message to the router telling "Ok, I'm done, you can remove it".

The problem occurs because at the moment the conference is considered ended, the channel remains open (the conference can't close because the conference only holds a receiver part of the channel). And in go, there is no way to check if the channel is alive or if there is someone listening on the channel. So this means that the router might have sent certain messages to the conference expecting that the conference will read them whereas in reality, the conference has stopped listening on the channel by the moment the router gets to know that the conference is dead.

To sum up:
- There is a way in Go to check if the channel is closed by the sender on the receiver's side.
- There is no way in Go to check if there is any listener on the sender's side.
- We need a way to know (on the sender side) if someone is listening to the channel (in order to know if it makes sense to post a message).

That's pretty much the problem that I tried to solve with this logic, i.e. create a wrapper around the channel where I could inform the sender that the receiver is not listening to the changes, so that if the sender tries to send the message to such a channel, it gets the message back and knows that there are no receivers anymore.

Reply:
Thanks for the detailed explanation - I've also found Go is a bit light on actually explaining how you would write correct & non-racy code sticking to the given paradigms.

Reply:
Right! I've also spent some time trying to understand how to do things safely. In Rust, for instance, we always know if there is a sender and if there is a receiver, but in Go, an elegant implementation of this does not work because of garbage collection: the receivers are not deleted immediately after leaving the scope of a function, but only get removed when the GC kicks in which means that the lifetime of the objects is not strictly defined, hence the consequence of us not knowing when the other counterpart of a channel is dead.